### ISSUE TITLE
[Proj0] What syscall number is better for `sys_hello`: 399 or 400?

### ISSUE THREAD

**Issue by gyusang** [swsnu/osspr2022#11]
As discussed in #10 the 399 syscalls in [unistd32.h](https://github.com/swsnu/tizen-6.0-rpi3/blob/380810fb47a85ddb59faee8875b8fa26ac8d57b0/arch/arm64/include/asm/unistd32.h#L820) exists as a compatibility layer for 32-bit applications running on 64-bit architecture.
Therefore, the syscall numbers in [`arch/arm64/.../unistd32.h`](https://github.com/swsnu/tizen-6.0-rpi3/blob/380810fb47a85ddb59faee8875b8fa26ac8d57b0/arch/arm64/include/asm/unistd32.h#L820) should match the syscall numbers in [`arch/arm/.../syscall.tbl`](https://github.com/swsnu/tizen-6.0-rpi3/blob/380810fb47a85ddb59faee8875b8fa26ac8d57b0/arch/arm/tools/syscall.tbl#L416) (the 32-bit *architecture* counterpart).

The project description suggests using syscall number 399 for `sys_hello`.
https://github.com/swsnu/osspr2022/blob/6d5cdfaec5a8f06e4a07afaec5bb639b647090bc/doc/Project0.md#L120

However, in [`arch/arm/.../syscall.tbl`](https://github.com/swsnu/tizen-6.0-rpi3/blob/380810fb47a85ddb59faee8875b8fa26ac8d57b0/arch/arm/tools/syscall.tbl#L416), the syscall number 399 is already used for `sys_io_pgetevents`. 
To provide full compatibility for 32-bit applications, syscall number 399 in [`arch/arm64/.../unistd32.h`](https://github.com/swsnu/tizen-6.0-rpi3/blob/380810fb47a85ddb59faee8875b8fa26ac8d57b0/arch/arm64/include/asm/unistd32.h#L820) should be assigned to `sys_io_pgetevents` (if not in the 4.19 kernel, in the future patches).

Then, would leaving syscall number 399 intact and using syscall number 400 for `sys_hello` more appropriate than using syscall number 399 for `sys_hello`?

**Comment by widiba03304** [comment]
I am replying this with my phone so this might not be sufficient information for the question. I will further explain this topic next afternoon! (I need sleep for now sorry)

io_pgetevents in arm64 is assigned to 292, and the compat layer is implemented properly. So there should be no conflict.

**Comment by widiba03304** [comment]
I don't know if this is right.

> Therefore, the syscall numbers in arch/arm64/.../unistd32.h should match the syscall numbers in arch/arm/.../syscall.tbl (the 32-bit architecture counterpart).

I disagree with this. Arm64 architecture does not have any duty to emulate all the features of other architecture. The compatibility layer is just a feature to provide some convenience to users.

Arm arch is not just a previous version of Arm64 arch. So, arm arch keeps changing though arm64 exists.

I think these can give some excuses for the mismatch you mentioned. To support this argument, I'll find some historical patch notes and let you know :)

**Comment by goranmoomin** [comment]
IMHO the cleanest way here would be just declaring `sys_hello` as a 64-bit only syscall and be done with compat problems (I eventually went down this route).

I think that there is an implicit expectation on the 32-bit compact layer that syscall numbers would match with the 32-bit kernel – My personal opinion is that syscalls that only exist on one side might make sense, but different syscalls on the same number are IMO hunting down for problems.

I briefly considered implementing `sys_io_pgetevents` as 399 on the compat layer – IIRC our provided kernel already has an implementation of `compat_sys_io_pgetevents`function so in theory I think it's just adding a line for it? But I didn't want to go through hoops trying to test `io_pgevents` just to implement `sys_hello`...

**Comment by widiba03304** [comment]
@goranmoomin I totally agree. I am planning to introduce this discussion in the practice session  which will be held in Mar 9.

Thank y'all for raising this concern!

**Comment by gyusang** [comment]
Since arm [maintains](https://developer.arm.com/tools-and-software/open-source-software/linux-kernel) the arch/arm64 linux kernel, the expectation that 32-bit compat layer syscall number matches the 32-bit kernel seems to be more than "implicit."
In the latest linux kernel source tree, almost all entries of [`syscall.tbl`](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm/tools/syscall.tbl) and [`unistd32.h`](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm64/include/asm/unistd32.h) matches, with some notable exceptions being old oabi syscalls not being implemented in arm64 `unistd32.h`.
This includes:
- `sys_syscall` (113) : [arm](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm/tools/syscall.tbl#L129) [arm64](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm64/include/asm/unistd32.h#L239-L240)
- `sys_socketcall` (102) : [arm](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm/tools/syscall.tbl#L117) [arm64](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm64/include/asm/unistd32.h#L217-L218)
- `sys_mmap` (90) : [arm](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm/tools/syscall.tbl#L105) [arm64](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm64/include/asm/unistd32.h#L193-L194)
- `sys_readdir` (89) : [arm](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm/tools/syscall.tbl#L104) [arm64](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm64/include/asm/unistd32.h#L191-L192)
- `sys_old_select` (82) : [arm](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm/tools/syscall.tbl#L97) [arm64](https://github.com/torvalds/linux/blob/dcde98da997075053041942ecf97d787855722ec/arch/arm64/include/asm/unistd32.h#L177-L178)
- ...

Even in these cases, the syscall number is deliberately skipped, by assigning the unimplemented syscall number to `sys_ni_syscall`.
```c
			/* 89 was sys_readdir */
__SYSCALL(89, sys_ni_syscall)
```

Therefore, if we were to implement `sys_hello` in the compat layer, IMO we should skip the number 399, either by explicitly stating it is reserved,
```c
			/* 399 reserved for sys_io_pgetevents */
__SYSCALL(399, sys_ni_syscall)
```
or by implicitly skipping the number and using 400, as the default for any syscall number in arm64 is `sys_ni_syscall` (See [sys32.c](https://github.com/swsnu/tizen-6.0-rpi3/blob/380810fb47a85ddb59faee8875b8fa26ac8d57b0/arch/arm64/kernel/sys32.c#L143-L149)).

**Comment by widiba03304** [comment]
Supporting certain syscalls with compatibility is determined by the maintainers. 

The concrete fact is that `sys_io_pgetevents` is implemented on arm, but it is not on arm64 and the maintainers decided not to support `sys_io_pgetevents` in this specific version.

Even in the v4.18 kernel for [arm](https://github.com/torvalds/linux/blob/v4.18/arch/arm/tools/syscall.tbl) and [arm64](https://github.com/torvalds/linux/blob/v4.18/arch/arm64/include/asm/unistd32.h) code, still the mismatch and some syscalls labeled to be not implemented (`sys_ni_syscall`) exists at the same time.

So my argument on this is that the mismatch exists on purpose. Maybe the compat layer of arm64 follows arm one step behind for some reason. (Because adding syscall should be done in careful and conservative way and add it to both is too speculative? idk...)

BTW, I think adding `sys_hello` to 64-bit syscall is the most proper way to tackle all of these. I will announce and explain in the practice session.

**Comment by goranmoomin** [comment]
> Even in the v4.18 kernel for [arm](https://github.com/torvalds/linux/blob/v4.18/arch/arm/tools/syscall.tbl) and [arm64](https://github.com/torvalds/linux/blob/v4.18/arch/arm64/include/asm/unistd32.h) code, still the mismatch and some syscalls labeled to be not implemented (sys_ni_syscall) exist at the same time.

AFAIK the ARM64 32-bit compat layer only supports the EABI variant, so it explicitly doesn't support syscalls that only exist in OABI.
